server/src/main/java/org/apache/iotdb/db/query/dataset/RawQueryDataSetWithValueFilter.java

```diff
@@ -73,26 +77,62 @@ protected RowRecord nextWithoutConstraint() throws IOException {
    * @return if there has next row record.
    */
   private boolean cacheRowRecord() throws IOException {
-    while (timeGenerator.hasNext()) {
+    Object[][] results = new Object[seriesReaderByTimestampList.size()][];
+    while (timeGenerator.hasNextTimeColumn()) {
+      initResults(results);
+
+      TimeColumn timeColumn = timeGenerator.nextTimeColumn();
+
+      fillResults(results, timeColumn);
+      flatAndCache(results, timeColumn);
+    }
+    hasCachedRowRecord = !cachedRecords.isEmpty();
+    return hasCachedRowRecord;
+  }
+
+  private void flatAndCache(Object[][] results, TimeColumn timeColumn) {
+    int resultSize = timeColumn.size() - timeColumn.position();
+    for (int i = 0; i < resultSize; i++) {
+      RowRecord rowRecord = new RowRecord(timeColumn.getTimeByIndex(i + timeColumn.position()));
       boolean hasField = false;
-      long timestamp = timeGenerator.next();
-      RowRecord rowRecord = new RowRecord(timestamp);
-      for (int i = 0; i < seriesReaderByTimestampList.size(); i++) {
-        IReaderByTimestamp reader = seriesReaderByTimestampList.get(i);
-        Object value = reader.getValueInTimestamp(timestamp);
-        if (value == null) {
+      for (Object[] result : results) {
+        if (result[i] == null) {
           rowRecord.addField(null);
-        } else {
-          hasField = true;
-          rowRecord.addField(value, dataTypes.get(i));
+          continue;
         }
+        hasField = true;
+        rowRecord.addField((Field) result[i]);
       }
       if (hasField) {
-        hasCachedRowRecord = true;
-        cachedRowRecord = rowRecord;
-        break;
+        cachedRecords.add(rowRecord);
       }
     }
-    return hasCachedRowRecord;
+  }
+
+  private void fillResults(Object[][] results, TimeColumn timeColumn) throws IOException {
```

Comment:
combine the fillResults, initReulsts, flatAndCache, cacheRowRecord into hasNextWithoutConstraints

Reply:
done